### ISSUE TITLE
JSQMessagesComposerTextView clears my menu Items

### ISSUE THREAD

**Issue by Slessi** [jessesquires/JSQMessagesViewController#1790]
I am adding custom menu items to my cells as follows:

```objective-c
- (void)viewDidLoad
{
    [super viewDidLoad];

    [JSQMessagesCollectionViewCell registerMenuAction:@selector(forward:)];
    [JSQMessagesCollectionViewCell registerMenuAction:@selector(saveImage:)];

    [UIMenuController sharedMenuController].menuItems = @[
        [[UIMenuItem alloc] initWithTitle:@"Forward" action:@selector(forward:)],
        [[UIMenuItem alloc] initWithTitle:@"Save" action:@selector(saveImage:)]
    ];
}
```

And then I add the necessary delegates etc. and the menu items show up fine. However after opening up the menu items for the `JSQMessagesComposerTextView`, my menu items are cleared! As seen [here](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/JSQMessagesViewController/Views/JSQMessagesComposerTextView.m#L295) on line 295.

Is this correct? Should I be setting my menu items in `collectionView:canPerformAction:forItemAtIndexPath:withSender:` ? It did fix my problem, but doesn't seem right to me

**Comment by Lucashuang0802** [comment]
We have included an example how to compose your custom menu items from `DemoMessagesViewController`. If you search for `Custom Action`, it will navigate you to the code: 
```
#pragma mark - Custom menu actions for cells

- (void)didReceiveMenuWillShowNotification:(NSNotification *)notification
{
    /**
     *  Display custom menu actions for cells.
     */
    UIMenuController *menu = [notification object];
    menu.menuItems = @[ [[UIMenuItem alloc] initWithTitle:@"Custom Action" action:@selector(customAction:)] ];
    
    [super didReceiveMenuWillShowNotification:notification];
}
```
`collectionView:canPerformAction:forItemAtIndexPath:withSender:` is just for filtering out some actions that are not needded to be there for some cases. It's for customization.

**Comment by Slessi** [comment]
Thank you :)

**Comment by Slessi** [comment]
@Lucashuang0802 @jessesquires 

> TL;DR
>
> If all my actions are custom actions, `didReceiveMenuWillShowNotification` never gets fired, so `shouldShowMenuForItemAtIndexPath` will always return `NO`.

Hi again,

I thought I had resolved this, but I believe I've found an actual bug this time.

In `viewDidLoad` I register a custom action called `forward`.
```objective-c
[JSQMessagesCollectionViewCell registerMenuAction:@selector(forward:)];
```

In `didReceiveMenuWillShowNotification:` I set the menu item for `forward`.
```objective-c
- (void)didReceiveMenuWillShowNotification:(NSNotification *)notification
{
    UIMenuController *menu = [notification object];
    menu.menuItems = @[
        [[UIMenuItem alloc] initWithTitle:@"Forward" action:@selector(forward:)]
    ];

    [super didReceiveMenuWillShowNotification:notification];
}
```

Then in my `canPerformAction:` I do the following:
```objective-c
- (BOOL)collectionView:(UICollectionView *)collectionView
      canPerformAction:(SEL)action
    forItemAtIndexPath:(NSIndexPath *)indexPath
            withSender:(id)sender
{
    if (action == @selector(copy:)) {
        return NO;
    }

    if (action == @selector(forward:)) {
        return YES;
    }

    return [super collectionView:collectionView
                canPerformAction:action
              forItemAtIndexPath:indexPath
                      withSender:sender];
}
```

You will note that I've set it to return `NO` for `@selector(copy:)`, i.e. none of the default actions will return `YES` here.

This means:
 1. `didReceiveMenuWillShowNotification` doesn't get called
 2. My custom menu item never gets set
 3. `canPerformAction:` will never be called for my custom action

I was able to reproduce this issue in the demo project.

I can resolve this if I set my `menuItems` twice, i.e. once in `viewDidLoad` and then again (as per demo) in `didReceiveMenuWillShowNotification`. Alternatively I can resolve this by going back to the initial solution I had attempted in my first post, i.e. setting the menu items in `collectionView:canPerformAction:forItemAtIndexPath:withSender:`.
